### PR TITLE
Arm backend: Minimal example of pruning

### PR DESCRIPTION
**Pull request by gggekov**
Explain how to prune a NN and the associated uplift in performance when running on the Ethos-U NPU.

cc @freddan80 @per @zingo @oscarandersson8218 @digantdesai

### CODE REVIEW

**Comment by zingo** [comment]
@gggekov the label "topic: not user facing" is not used anymore so no need to set it it was replaced with "release notes: arm"  or "release notes: none"

File: examples/arm/pruning_minimal_example.ipynb [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+    "    acc = 100 * correct / total\n",
+    "    print(f\"Top 1 accuracy = {acc:.2f}%\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a4750eaf",
+   "metadata": {},
+   "source": [
+    "Let's instantiate the model and train it. In order to get reproduceable results, we will fix the seed."
```
[the remark this answers is not in the document]
Comment (gggekov) [reply examples/arm/pruning_minimal_example.ipynb]: Done

File: examples/arm/pruning_minimal_example.ipynb [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+    "print(\"Evaluate FP32 model accuracy\")\n",
+    "evaluate(model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "9837d9ba",
+   "metadata": {},
+   "source": [
+    "We obtian 96% top1 accuracy for the FP32 model.\n",
```
[the remark this answers is not in the document]
Comment (gggekov) [reply examples/arm/pruning_minimal_example.ipynb]: Done.

File: examples/arm/pruning_minimal_example.ipynb [hunk 3]
```diff
@@ -0,0 +1,567 @@
...
+   "cell_type": "markdown",
+   "id": "2b6cae04",
+   "metadata": {},
+   "source": [
+    "Note that as part of the compilation process in `to_edge_transform_and_lower`, we get Weight Compression information:\n",
+    "```\n",
+    "Original Weights Size                          522.50 KiB\n",
+    "NPU Encoded Weights Size                       507.44 KiB\n",
+    "```\n",
+    "In other words, the original Weights are 522KB and after compilation and encoding by the compiler, we get 507KB of weights that will be read by the NPU at runtime. Remembmer this is for the case when we've not applied pruning or clustering. This will generate original_model.pte file that we will deploy on device later on. \n",
```
[the remark this answers is not in the document]
Comment (gggekov) [reply examples/arm/pruning_minimal_example.ipynb]: Done.

File: examples/arm/pruning_minimal_example.ipynb [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import torch\n",
+    "from torchvision import datasets, transforms\n",
+    "from torch.utils.data import DataLoader\n",
+    "from torch import nn\n",
+    "import torch.nn.utils.prune as prune\n",
+    "import torch.nn.functional as F\n",
+    "from torch.utils.data import DataLoader, Subset\n",
```
[the remark this answers is not in the document]
Comment (gggekov) [reply examples/arm/pruning_minimal_example.ipynb:57]: Done.

File: examples/arm/pruning_minimal_example.ipynb [hunk 5]
```diff
@@ -0,0 +1,567 @@
...
+    "../../backends/arm/scripts/run_fvp.sh --elf=ethos_u_pruned_model/arm_executor_runner --target=ethos-u85-128"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e55ae929",
+   "metadata": {},
+   "source": [
+    "On the pruned model, the inference completes in 22k NPU cycles. The NPU still performs 8k MACs, but this time the number of cycles when the weight decoder is active has dropped to to 17k cycles. \n",
+    "It's also worth noting that the size of the pte file has been reducded significantly - from 518 KB of the original model to 57KB of the pruned workload. "
```
[the remark this answers is not in the document]
Comment (gggekov) [reply examples/arm/pruning_minimal_example.ipynb]: Done.

File: examples/arm/pruning_minimal_example.ipynb [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d934fe41",
+   "metadata": {},
+   "source": [
+    "# Conclusion\n",
+    "We defined a simple model to solve the MNIST dataset. The model is using Linear layers and is heavily memory-bound on the external memory. We pruned the model and obtain similar int8 accuracy between the original workload and the pruned counterpart. Let us put the results from the runtime in a table and draw a few conclusions: \n",
+    "\n",
+    "| Model                                   |NPU_ACTIVE cycles | NPU Encoded Weight Size   | Weight Decoder Active Cycles |    External memory beats read   | Size of the pte file  |\n",
```
[the remark this answers is not in the document]
Comment (gggekov) [reply examples/arm/pruning_minimal_example.ipynb:536]: Beats is correct - the PMU counter in question reports how many beats are read.

File: examples/arm/executor_runner/arm_perf_monitor.cpp [hunk 7]
```diff
@@ -65,11 +65,14 @@ void ethosu_inference_begin(struct ethosu_driver* drv, void*) {
   ETHOSU_PMU_Set_EVTYPER(drv, 2, ETHOSU_PMU_EXT_RD_DATA_BEAT_RECEIVED);
   ETHOSU_PMU_Set_EVTYPER(drv, 3, ETHOSU_PMU_EXT_WR_DATA_BEAT_WRITTEN);
   ETHOSU_PMU_Set_EVTYPER(drv, 4, ETHOSU_PMU_NPU_IDLE);
+  ETHOSU_PMU_Set_EVTYPER(drv, 5, ETHOSU_PMU_MAC_ACTIVE);
+  ETHOSU_PMU_Set_EVTYPER(drv, 6, ETHOSU_PMU_WD_ACTIVE);
   // Enable the 5 counters
```
[the remark this answers is not in the document]
Comment (gggekov) [reply examples/arm/executor_runner/arm_perf_monitor.cpp]: Done.

**Review by zingo — APPROVED** [review]

**Comment by zingo** [comment]
Fails are unrelated
